lib/actions-oclif/preload.ts

```diff
@@ -0,0 +1,549 @@
+
+		const spinnerHandler = function (event: any) {
+			let spinner = spinners[event.name];
+			if (!spinner) {
+				spinner = spinners[event.name] = new visuals.Spinner(event.name);
+			}
+			if (event.action === 'start') {
+				return spinner.start();
+			} else {
+				console.log();
+				return spinner.stop();
+			}
+		};
+
+		options.commit = this.isCurrentCommit(options.commit || '')
+			? 'latest'
+			: options.commit;
+		options.image = params.image;
+		options.appId = options.app;
+		delete options.app;
+
+		options.splashImage = options['splash-image'];
+		delete options['splash-image'];
+
+		options.dontCheckArch = options['dont-check-arch'] || false;
+		delete options['dont-check-arch'];
+		if (options.dontCheckArch && !options.appId) {
+			throw new ExpectedError(
+				'You need to specify an app id if you disable the architecture check.',
+			);
+		}
+
+		options.pinDevice = options['pin-device-to-release'] || false;
+		delete options['pin-device-to-release'];
+
+		let certificates: string[];
+		if (Array.isArray(options['add-certificate'])) {
+			certificates = options['add-certificate'];
+		} else if (options['add-certificate'] === undefined) {
+			certificates = [];
+		} else {
+			certificates = [options['add-certificate']];
+		}
+		for (const certificate of certificates) {
+			if (!certificate.endsWith('.crt')) {
+				throw new ExpectedError('Certificate file name must end with ".crt"');
+			}
+		}
+
+		// Get a configured dockerode instance
+		const dockerUtils = await import('../utils/docker');
+		const docker = await dockerUtils.getDocker(options);
+		const preloader = new balenaPreload.Preloader(
+			null,
+			docker,
+			options.appId,
+			options.commit,
+			options.image,
+			options.splashImage,
+			options.proxy, // TODO: Currently always undefined, investigate approach in ssh command.
+			options.dontCheckArch,
+			options.pinDevice,
+			certificates,
+		);
+
+		let gotSignal = false;
+
+		nodeCleanup(function (_exitCode, signal) {
+			if (signal) {
+				gotSignal = true;
+				nodeCleanup.uninstall(); // don't call cleanup handler again
+				preloader.cleanup().then(() => {
+					// calling process.exit() won't inform parent process of signal
+					process.kill(process.pid, signal);
+				});
+				return false;
+			}
+		});
+
+		if (process.env.DEBUG) {
+			preloader.stderr.pipe(process.stderr);
+		}
+
+		preloader.on('progress', progressHandler);
+		preloader.on('spinner', spinnerHandler);
+
+		try {
+			await new Promise((resolve, reject) => {
+				preloader.on('error', reject);
+				resolve(this.prepareAndPreload(preloader, balena, options));
+			});
+		} catch (err) {
+			if (instanceOf(err, balena.errors.BalenaError)) {
+				const code = err.code ? `(${err.code})` : '';
+				throw new ExpectedError(`${err.message} ${code}`);
+			} else {
+				throw err;
+			}
+		} finally {
+			if (!gotSignal) {
+				await preloader.cleanup();
+			}
+		}
+	}
+
+	readonly applicationExpandOptions: any = {
```

Comment:
You should be able to type this as
```suggestion
	readonly applicationExpandOptions: PinejsClientCore.ResourceExpand = {
```

Reply:
Do you know where I can find that namespace/type?  IDE not helping.

Reply:
It's in the `pinejs-client-core` module

Reply:
Although balena-sdk uses some typings customized to the specific api model so it might be possible to use that - @thgreasi would know how to do that

Reply:
Hmm, Looks like that `ResourceExpand` type is generic.  If I use `ResourceExpand<any>` I get :
```
Type '{ end_timestamp: string; }' is not assignable to type 'OrderByValues'.   Type '{ end_timestamp: string; }' is not assignable to type '"desc"'.
```